chain/client/src/sync/state.rs

```diff
@@ -993,25 +1015,78 @@ impl StateSync {
         sync_hash: CryptoHash,
         chain: &mut Chain,
-        now: Utc,
+        load_memtrie_scheduler: &near_async::messaging::Sender<LoadMemtrieRequest>,
     ) -> Result<(), near_chain::Error> {
         // Keep waiting until our shard is on the list of results
         // (these are set via callback from ClientActor - both for sync and catchup).
         if let Some(result) = self.state_parts_apply_results.remove(&shard_id) {
-            match chain.set_state_finalize(shard_id, sync_hash, result) {
-                Ok(()) => {
-                    *shard_sync_download = ShardSyncDownload {
-                        downloads: vec![],
-                        status: ShardSyncStatus::StateDownloadComplete,
+            result?;
+            let epoch_id = chain.get_block_header(&sync_hash)?.epoch_id().clone();
+            let shard_uid = chain.epoch_manager.shard_id_to_uid(shard_id, &epoch_id)?;
+            let shard_state_header = chain.get_state_header(shard_id, sync_hash)?;
+            let chunk = shard_state_header.cloned_chunk();
+            let block_hash = chunk.prev_block();
+
+            // We synced shard state on top of _previous_ block for chunk in shard state header and applied state parts to
+            // flat storage. Now we can set flat head to hash of this block and create flat storage.
+            // If block_hash is equal to default - this means that we're all the way back at genesis.
+            // So we don't have to add the storage state for shard in such case.
+            // TODO(8438) - add additional test scenarios for this case.
+            if *block_hash != CryptoHash::default() {
+                chain.create_flat_storage_for_shard(shard_uid, &chunk)?;
+            }
+            // We schedule load memtrie when flat storage state (if any) is ready.
+            // It is possible that memtrie is not enabled for that shard,
+            // in which case the task would finish immediately with Ok() status.
+            // We require the task result to further proceed with state sync.
+            chain.schedule_load_memtrie(shard_uid, sync_hash, &chunk, load_memtrie_scheduler);
+            *shard_sync_download = ShardSyncDownload {
+                downloads: vec![],
+                status: ShardSyncStatus::StateApplyFinalizing,
+            }
+        }
+        Ok(())
+    }
+
+    fn sync_shards_apply_finalizing_status(
+        &mut self,
+        shard_id: ShardId,
+        chain: &mut Chain,
+        sync_hash: CryptoHash,
+        now: Utc,
+        need_to_reshard: bool,
+        shard_sync_download: &mut ShardSyncDownload,
+    ) -> Result<bool, near_chain::Error> {
+        // Keep waiting until our shard is on the list of results
+        // (these are set via callback from ClientActor - both for sync and catchup).
+        let mut shard_sync_done = false;
+        if let Some(result) = self.load_memtrie_results.remove(&shard_id) {
+            result
```

Comment:
nit: Why the `and_then` and `or_else` here? It's slightly hard to read. Can't we do a match instead?

Reply:
We want the `or_else` arm both if result is error or `set_state_finalize()` returns error. If I simply do match, I would have to duplicate the `or_else` arm code. I did not find a cleaner approach.